File: Array/medium/cpp/sort_an_array_of_0s_1s_and_2s.cpp
//
// Created by santosh on 12/19/25.
//


#include<bits/stdc++.h>
using namespace std;

class Solution
{
    public:
    void selectionSort(vector<int>& nums)
    {

        for (int i =0; i<nums.size(); i++)
        {
            for (int j=i+1; j<nums.size(); j++)
            {
                if (nums[i] > nums[j])
                {
                    swap(nums[i], nums[j]);
                }
            }
        }
    }


    void dutchNationalFlagAlgorithm(vector<int>& nums)
    {



        int left = 0;
        int high = nums.size()-1;
        int mid= 0;

        while (mid<=high)
        {
            if (nums[mid] ==0)
            {
                swap(nums[left],nums[mid]);
                left++;
                mid++;
            }
            else if (nums[mid] ==1)
            {
                mid++;
            }
            else if (nums[mid] ==2)
            {
                swap(nums[mid],nums[high]);
                high--;
            }
        }
    }
};


int main()
{
    Solution s;
    vector<int> nums={2,0,2,1,1,0};

    s.dutchNationalFlagAlgorithm(nums);
    for (int i=0; i<nums.size(); i++)
    {
        cout<<nums[i]<<" ";
    }
}

File: Array/medium/python/sort_an_array_of_0s_1s_and_2s.py
# Sort an array of 0s, 1s and 2s (Dutch National Flag Algorithm) (LeetCode 75)
# Created by santosh

class Solution:
    def selection_sort(self, nums: list[int]) -> None:
        """Selection sort O(n^2)"""
        n = len(nums)
        for i in range(n):
            for j in range(i + 1, n):
                if nums[i] > nums[j]:
                    nums[i], nums[j] = nums[j], nums[i]

    def dutch_national_flag_algorithm(self, nums: list[int]) -> None:
        """
        Dutch National Flag Algorithm (Three-way partition).
        Time: O(n), Space: O(1)
        """
        left = 0
        high = len(nums) - 1
        mid = 0

        while mid <= high:
            if nums[mid] == 0:
                nums[left], nums[mid] = nums[mid], nums[left]
                left += 1
                mid += 1
            elif nums[mid] == 1:
                mid += 1
            elif nums[mid] == 2:
                nums[mid], nums[high] = nums[high], nums[mid]
                high -= 1


if __name__ == "__main__":
    s = Solution()
    nums = [2, 0, 2, 1, 1, 0]
    s.dutch_national_flag_algorithm(nums)
    print(nums)
